Editor/AGS.Editor/BuildTargets/BuildTargetDataFile.cs

```diff
@@ -59,7 +64,78 @@ private string[] ConstructFileListForDataFile()
             }
             Utilities.AddAllMatchingFiles(files, "*.ogv");
-            return files.ToArray();
+
+            // Regular files are registered under their filenames (w/o dir)
+            var assetList = new List<Tuple<string, string>>();
+            foreach (var f in files)
+            {
+                assetList.Add(new Tuple<string, string>(Path.GetFileName(f), f));
+            }
+
+            var userFiles = ConstructCustomFileListForDataFile(errors);
+            if (userFiles != null) assetList.AddRange(userFiles);
+            return assetList.ToArray();
+        }
+
+        /// <summary>
+        /// Creates a list of custom game resources as a list of tuples:
+        /// - first tuple's element is resource's name,
+        /// - second is real file path
+        /// </summary>
+        private Tuple<string, string>[] ConstructCustomFileListForDataFile(CompileMessages errors)
+        {
+            string userOpt = Factory.AGSEditor.CurrentGame.Settings.CustomDataDir;
+            if (string.IsNullOrEmpty(userOpt)) return null;
+
+            string curDir = Factory.AGSEditor.CurrentGame.DirectoryPath;
+            string[] restrictedDirs = new string[Factory.AGSEditor.RestrictedGameDirectories.Length];
+            for (int i = 0; i < restrictedDirs.Length; ++i)
+                restrictedDirs[i] = Path.Combine(curDir, Factory.AGSEditor.RestrictedGameDirectories[i]);
+
+            List<string> userFiles = new List<string>();
+            string[] user_dirs = userOpt.Split(',');
+            Array.Sort(user_dirs);
+            List<string> usedCustomDirs = new List<string>();
+
+            foreach (var dir in user_dirs)
+            {
+                // don't allow absolute paths
+                if (Path.IsPathRooted(dir))
+                {
+                    errors.Add(new CompileWarning(
+                        string.Format("Cannot use absolute path as a custom data source: {0}", dir)));
+                    continue;
+                }
+                // don't allow paths with "." or ".." in them
+                if (Utilities.DoesPathContainDotDirs(dir))
+                {
+                    errors.Add(new CompileWarning(
+                        string.Format("Cannot use path containing '.' or '..' as a custom data source: {0}", dir)));
+                    continue;
+                }
+                // don't allow paths that match or inside restricted folders,
+                // or folders that were already used to make a list of assets
+                string testDir = Path.Combine(curDir, dir);
+                if (Utilities.PathIsSameOrNestedAmong(testDir, restrictedDirs))
+                {
+                    errors.Add(new CompileWarning(
+                        string.Format("Cannot use restricted location as a custom data source: {0}", dir)));
+                    continue;
+                }
+                if (Utilities.PathIsSameOrNestedAmong(testDir, usedCustomDirs.ToArray()))

```

Comment:
If `Utilities.PathIsSameOrNestedAmong` input is changed so it accepts IEnumerable<string> as second parameter you can pass any collection (array, list, queue, etc.) without having to reallocate them to another data structure.